engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnEventSubscriptionBehavior.java

```diff
@@ -0,0 +1,199 @@
+    eventRecord.setElementId(eventTrigger.getElementId());
+    eventRecord.setBpmnElementType(BpmnElementType.BOUNDARY_EVENT);
+
+    final var boundaryEvent = getBoundaryEvent(element, context, eventTrigger);
+
+    final long boundaryElementInstanceKey = keyGenerator.nextKey();
+    if (boundaryEvent.interrupting()) {
+
+      deferBoundaryEvent(context, boundaryElementInstanceKey);
+
+      stateTransitionBehavior.transitionToTerminating(context);
+
+    } else {
+      activateBoundaryEvent(context, boundaryElementInstanceKey, eventRecord);
+    }
+
+    stateBehavior
+        .getVariablesState()
+        .setTemporaryVariables(boundaryElementInstanceKey, eventTrigger.getVariables());
+
+    eventScopeInstanceState.deleteTrigger(
+        context.getElementInstanceKey(), eventTrigger.getEventKey());
+  }
+
+  private <T extends ExecutableActivity> ExecutableBoundaryEvent getBoundaryEvent(
+      final T element, final BpmnElementContext context, final EventTrigger eventTrigger) {
+
+    return element.getBoundaryEvents().stream()
+        .filter(boundaryEvent -> boundaryEvent.getId().equals(eventTrigger.getElementId()))
+        .findFirst()
+        .orElseThrow(
+            () ->
+                new IllegalStateException(
+                    String.format(
+                        "Expected boundary event with id '%s' but not found. [context: %s]",
+                        BufferUtil.bufferAsString(eventTrigger.getElementId()), context)));
+  }
+
+  private void deferBoundaryEvent(
+      final BpmnElementContext context, final long boundaryElementInstanceKey) {
+
+    elementInstanceState.storeRecord(
+        boundaryElementInstanceKey,
+        context.getElementInstanceKey(),
+        eventRecord,
+        WorkflowInstanceIntent.ELEMENT_ACTIVATING,
+        Purpose.DEFERRED);
+  }
+
+  public void publishTriggeredBoundaryEvent(final BpmnElementContext context) {
+
+    elementInstanceState.getDeferredRecords(context.getElementInstanceKey()).stream()
+        .filter(record -> record.getValue().getBpmnElementType() == BpmnElementType.BOUNDARY_EVENT)
+        .findFirst()
+        .ifPresent(
+            deferredRecord ->
+                activateBoundaryEvent(context, deferredRecord.getKey(), deferredRecord.getValue()));
+  }
+
+  private void activateBoundaryEvent(
+      final BpmnElementContext context,
+      final long elementInstanceKey,
+      final WorkflowInstanceRecord eventRecord) {
+
+    streamWriter.appendNewEvent(
+        elementInstanceKey, WorkflowInstanceIntent.ELEMENT_ACTIVATING, eventRecord);
+
+    stateBehavior.createElementInstanceInFlowScope(context, elementInstanceKey, eventRecord);
+    stateBehavior.spawnToken(context);
+  }
+
+  public <T extends ExecutableCatchEventSupplier> Either<Failure, Void> subscribeToEvents(
+      final T element, final BpmnElementContext context) {
+
+    try {
+      catchEventBehavior.subscribeToEvents(context.toStepContext(), element);
+      return Either.right(null);
```

Comment:
What is the benefit here to use this either?

Reply:
It allows chaining the method calls. See here:
```
variableMappingBehavior
        .applyInputMappings(context, element)
        .flatMap(ok -> eventSubscriptionBehavior.subscribeToEvents(element, context))
        .ifRightOrLeft(
            ok -> stateTransitionBehavior.transitionToActivated(context),
            failure -> incidentBehavior.createIncident(failure, context));
```

Without the `Either`, this logic is a mess. We need to create an incident is something bad happens while applying the input mappings or subscribing to the events. If everything is ok then we can take the transition to the next state.

Reply:
thanks